Sprint Zug PHX-201 Allow HTML to be included in XBlock Field help strings

douglashall opened the pull request:
I am migrating the LTI component from https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/lti_module.py to a new XBlock repo and I would like to use this package to render the studio edit view of the XBlock.

The existing xmodule implementation includes some HTML links to documentation in the help attribute of Field definitions. This change will allow those HTML links to be displayed when the studio edit view is rendered.

The existing xmodule implementation of the LTI component also uses Mako templates. I have added `ResourceLoader.render_mako_template` in order to support this. I have also added `ResourceLoader.render_django_template` and deprecated `ResourceLoader.render_template` in order to keep the API naming consistent.

douglashall commented:
@antoviaque Can you take a look at this one for me?

antoviaque commented:
@douglashall Thanks for the PR! : ) On a quick look, the change makes sense to me, and doesn't seem to represent a security risk, given that the value would always be defined from the code (or the translation files), as far as I can tell. 

Would you be able to add a test for this, to ensure this isn't broken inadvertently in the future?

I'll also still schedule a formal review for next week to get a second pair of eyes on it. Maybe at the same time as a review on the LTI XBlock? Will you need a review from us there, and if so when do you expect it to be up?

douglashall commented:
@antoviaque Sure thing. I will add a test. I am on schedule to have the LTI XBlock ready for review by early next week.

antoviaque commented:
@douglashall Perfect - I'll put both reviews on our list for next week then.

Thanks for the work on the LTI XBlock btw - it's great to see activity on it : )

smarnach commented on xblockutils/resources.py:
```diff
@@ -85,3 +88,14 @@ def load_scenarios_from_path(self, relative_scenario_dir, include_identifier=Fal
 
         return scenarios
+
+class MakoResourceLoader(ResourceLoader):
```
This doesn't really feel like a separate class to me.  Django and Mako templates aren't mutually exclusive in any way, so how about just adding a new method `render_mako_template()` to `ResourceLoader`?  (We could also deprecate the method name `render_template()` and call it `render_django_template()` instead for consistency, if you like that better.)

douglashall replied:
Apologies @smarnach, I missed this comment before. I have now added ResourceLoader.render_mako_template and ResourceLoader.render_django_template and deprecated ResourceLoader.render_template.

smarnach commented on xblockutils/templates/studio_edit.html:
```diff
@@ -93,7 +93,7 @@
           </div>
           {% if field.help %}
-            <span class="tip setting-help"> {{ field.help }} </span>
+            <span class="tip setting-help"> {{ field.help|safe }} </span>
```
This change isn't fully backwards compatible.  Some existing help string might rely on special characters being escaped, though that seems unlikely in practice.  I can't think of an easy way to verify this for all XBlocks using this library, so I'm fine with the change.

smarnach commented:
@douglashall These changes make sense to me.  Could you please add a unit test for the Mako template renderer (and also the test Xavier asked for)?  It might also make sense to update the PR description.

douglashall commented:
@smarnach @antoviaque Could you take another look at this when you get a chance? I have added unit tests for the new functionality and updated the PR description.

smarnach commented:
@douglashall Looks good to me.  Did you see [this comment above](https://github.com/edx/xblock-utils/pull/32#discussion_r44426404)?

smarnach commented:
@douglashall Thanks for the updates.  Looks all good and tests are passing, :+1:.

douglashall commented:
@chrisndodge Can you give this a second review when you get a chance? Thanks.

chrisndodge commented:
@douglashall I see Xavier and Sven already did a good 1st review. My recommendation would be to get someone from the Platform team to act as 2nd reviewer, like @cpennington or @doctoryes. That said, I'm going to download the main PR in edx-platform and try it out locally from an end-user (student and author) perspective.

douglashall commented:
@doctoryes @macdiesel Could I get a second review on this when someone gets a chance?

douglashall commented:
@macdiesel Can you take a look at this PR when you get a chance?

doctoryes commented on tests/integration/test_studio_editable.py:
```diff
@@ -274,13 +275,17 @@ class FancyXBlock(StudioEditableXBlockMixin, XBlock):
         'int_values', 'list_normal', 'list_intdefs', 'list_strdefs', 'list_set_ints', 'list_set_strings',
         'string_normal', 'string_values', 'string_values_provider', 'string_named', 'string_dynamic',
-        'string_multiline', 'string_multiline_reset', 'string_html',
+        'string_multiline', 'string_multiline_reset', 'string_html', 'string_with_help'
```
The `README` talks about `string_multiline` and `string_html` but doesn't have info about `string_multiline_reset` or the new `string_with_help`. Is a `README` update appropriate?

douglashall replied:
Yes. I will update. Thanks.

douglashall replied:
`string_multiline_reset` and `string_with_help` are fields on this test XBlock class that exercise the existing `allow_reset` and `help` Field kwargs. Those kwargs don't offer any new functionality above what the XBlock API offers. I believe String (multiline) and String (html) are documented in the README because they illustrate that this package adds the `multiline_editor` Field kwarg which can be used to control the display of fields in Studio.

doctoryes commented:
One documentation nit. Otherwise, nice work. :+1:

macdiesel commented:
:+1:  good stuff
